modules/account-lib/src/mpc/shamir.ts

```diff
@@ -0,0 +1,98 @@
+const assert = require('assert');
+
+const Shamir = (curve: any) => {
+  /**
+   * Perform Shamir sharing on the secret `secret` to the degree `threshold - 1` split `numShares`
+   * ways. The split secret requires `threshold` shares to be reconstructed.
+   *
+   * @param secret secret to split
+   * @param threshold share threshold required to reconstruct secret
+   * @param numShares total number of shares to split to split secret into
+   * @param indices
+   * @returns Dictionary of shares. Each key is an int in the range 1<=x<=numShares
+   * representing that share's free term.
+   */
+  const split = (
+    secret: Buffer,
+    threshold: number,
+    numShares: number,
+    indices?: Array<number>,
+  ): Record<number, Buffer> => {
+    if (indices === undefined) {
+      // make range(1, n + 1)
+      indices = Array.from({ length: numShares }, (_, i) => i + 1);
+    }
+    assert(threshold > 1);
+    assert(threshold <= numShares);
+    const coefs: Buffer[] = [];
+    for (let ind = 0; ind < threshold - 1; ind++) {
+      const random_value = curve.scalarRandom();
+      coefs.push(random_value);
+    }
+    coefs.push(secret);
+
+    const shares: Record<number, Buffer> = {};
+    for (let ind = 0; ind < indices.length; ind++) {
+      const x = indices[ind];
+      const x_buffer = Buffer.alloc(32);
+      x_buffer.writeUInt32LE(x, 0);
+      let partial = coefs[0];
+      for (let other = 1; other < coefs.length; other++) {
+        const scalarMult = curve.scalarMult(partial, x_buffer);
+        const newAdd = curve.scalarAdd(coefs[other], scalarMult);
+        partial = newAdd;
+      }
+      shares[x] = Buffer.from(partial);
+    }
+    return shares;
+  };
+
+  /**
+   * Reconstitute a secret from a dictionary of shares. The number of shares must
+   * be equal to `t` to reconstitute the original secret.
+   *
+   * @param shares dictionary of shares. each key is the free term of the share
+   * @returns secret
+   */
+  const combine = (shares: Record<number, Buffer>): Buffer => {
+    let s = Buffer.alloc(32);
+    for (const xi in shares) {
+      const yi = shares[xi];
+      const xi_buffer = Buffer.alloc(32);
+      let num_buffer = Buffer.alloc(32);
+      let denum_buffer = Buffer.alloc(32);
+      xi_buffer.writeUInt32LE(parseInt(xi, 10), 0);
+      num_buffer.writeUInt32LE(1, 0);
+      denum_buffer.writeUInt32LE(1, 0);
+
+      for (const xj in shares) {
+        const xj_buffer = Buffer.alloc(32);
+        xj_buffer.writeUInt32LE(parseInt(xj, 10), 0);
+        if (xi !== xj) {
+          num_buffer = curve.scalarMult(num_buffer, xj_buffer);
+        }
```

Comment:
This seems a little precarious. Since you're using buffers as your internal value type, you need to assume endianess, and here you're assuming little endian by writing `xj` to the zeroth index. This will work for Ed25519 since you'll be using sodium for the math, and sodium expects little endian buffers. However, this presents an opportunity to shoot yourself in the foot when dealing with secp256k1 which expects buffers to be big endian. You could solve for this by converting buffers from little to big endian in the secp256k1 curve implementation, but that seems a little hacky. The better solution would be to use native BigInt as the internal type and have the curve implementation convert to buffers themselves.

Reply:
converted to numbers as internal buffers and ran into some errors. added a task and todo to shamir for conversion to internal number type representation for when we add `secp256k1`. for now, figured to unblock other people to use this thought should be OK to still have the buffer representation